src/psyclone/psyir/backend/fortran.py

```diff
@@ -239,26 +240,65 @@ class FortranWriter(PSyIRVisitor):
+        the original.)
 
-        :param symbol: the symbol instance.
-        :type symbol: :py:class:`psyclone.psyir.symbols.DataSymbol`
+        :param symbol: the container symbol instance.
+        :type symbol: :py:class:`psyclone.psyir.symbols.ContainerSymbol`
 
-        :returns: the Fortran use statement as a string.
+        :returns: the Fortran use statement(s) as a string.
         :rtype: str
 
-        :raises VisitorError: if the symbol argument does not specify \
-        a use statement (its interface value is not a Global instance).
-
+        :raises VisitorError: if the symbol argument is not a ContainerSymbol.
+        :raises VisitorError: if a symbol listed as being imported from the \
+                    supplied container symbol does not have a GlobalInterface.
+        :raises VisitorError: if a symbol listed as being imported from the \
+                    supplied container has a GlobalInterface that references \
+                    a different container.
+        :raises VisitorError: if we fail to find a list of symbols to import \
+                    from the supplied container symbol and it is not marked \
+                    as having a wildcard import.
         '''
-        if not symbol.is_global:
-            raise VisitorError(
-                "gen_use() requires the symbol interface for symbol '{0}' to "
-                "be a Global instance but found '{1}'."
-                "".format(symbol.name, type(symbol.interface).__name__))
-
-        return "{0}use {1}, only : {2}\n".format(
-            self._nindent, symbol.interface.container_symbol.name, symbol.name)
+        if not isinstance(symbol, ContainerSymbol):
+            raise VisitorError("gen_use() expects a ContainerSymbol but got "
+                               "'{0}'".format(type(symbol).__name__))
+        # Construct the list of symbol names for the ONLY clause and perform
+        # consistency checks.
+        only_list = []
+        for dsym in symbol.imported_symbols:
+            if not dsym.is_global:
+                raise VisitorError(
+                    "Symbol '{0}' is marked as being imported from container "
+                    "'{1}' but has an interface of '{2}' rather than the "
+                    "expected Global interface.".format(
+                        dsym.name, symbol.name, type(dsym.interface).__name__))
+            if dsym.interface.container_symbol is not symbol:
+                raise VisitorError(
+                    "Container '{0}' lists symbol '{1}' among its imports but "
+                    "the global interface of that symbol refers to a different"
+                    " container ('{2}').".format(
+                        symbol.name, dsym.name,
+                        dsym.interface.container_symbol.name))
+            only_list.append(dsym.name)
+        # Another consistency check
+        if not only_list and not symbol.has_wildcard_import:
```

Comment:
I've googled but to no avail. Please could you remind me of the purpose of such a thing? (Is it something to do with keeping things in scope so that they don't get garbage collected?)

Reply:
Yes, I think so. If a module variable goes out of scope without having the save attribute (or being initialised which implies save apparently) then its value is undefined. Having said that, some compilers save the values anyway but it should not be assumed that they will. If we are not going to support this then the front end should reject it, not the back end. Personally I think we should support it and add a note in the PSyIR documentation explaining what it means.

Reply:
Yes, I agree. I will "make it so"...

Reply:
In principle, it would be valid Fortran to have:

    module some_mod
      use my_mod, only:
      use my_mod
    end module some_mod

Since neither of these `USE`s have an associated list of symbols, in the current implementation we have no way of distinguishing that case from a single `use my_mod` (wildcard import). Practically speaking, if a piece of code does have both `USE my_mod, ONLY:` *and* `USE my_mod` then the latter effectively renders the former redundant. In fact, *any* import of data from `my_mod` renders the use of `USE my_mod, ONLY:` redundant. We therefore have two choices:

  1) Do not explicitly capture the fact that there is an ONLY with an empty list. This means that
  any other USEs of that module will result in the USE with an empty ONLY list being omitted
  from the generated Fortran. This is functionally correct.
  2) Explicitly capture the presence of an ONLY with an empty list. This means that the Fortran
  backend can generate syntactically identical code to that found by the frontend.

So far, we've always designed things such that no information is lost when converting from Fortran into the PSyIR. This favours option 2. However, I'm not certain this is the best approach here since all we're doing is retaining information that the original Fortran was non-sensical.

Reply:
After discussion in the office, we've decided that option 1. is best as there's no point polluting the PSyIR with redundancy.